code/binary-tree/ViewModel/src/main/java/ru/unn/agile/binarytree/viewmodel/ViewModel.java

```diff
@@ -0,0 +1,218 @@
+
+    private final StringProperty addKey = new SimpleStringProperty();
+    private final StringProperty addValue = new SimpleStringProperty();
+    private final StringProperty findKey = new SimpleStringProperty();
+    private final StringProperty removeKey = new SimpleStringProperty();
+
+    private final StringProperty addStatus = new SimpleStringProperty();
+
+    private final StringProperty findResult = new SimpleStringProperty();
+    private final StringProperty findStatus = new SimpleStringProperty();
+
+    private final StringProperty removeResult = new SimpleStringProperty();
+    private final StringProperty removeStatus = new SimpleStringProperty();
+
+    public ViewModel() {
+        addKey.set("");
+        addValue.set("");
+        addStatus.set(Status.WAITING.toString());
+
+        findKey.set("");
+        findResult.set("");
+        findStatus.set(Status.WAITING.toString());
+
+        removeKey.set("");
+        removeResult.set("");
+        removeStatus.set(Status.WAITING.toString());
+
+        // Add listeners to the input text fields
+        final List<StringProperty> fields = new ArrayList<>() {
+            {
+                add(addKey);
+                add(addValue);
+                add(findKey);
+                add(removeKey);
+            }
+        };
+
+        for (StringProperty field : fields) {
+            final ValueChangeListener listener = new ValueChangeListener();
+            field.addListener(listener);
+            valueChangedListeners.add(listener);
+        }
+    }
+
+    public StringProperty addKeyProperty() {
+        return addKey;
+    }
+
+    public StringProperty addValueProperty() {
+        return addValue;
+    }
+
+    public StringProperty addStatusProperty() {
+        return addStatus;
+    }
+
+    public StringProperty findKeyProperty() {
+        return findKey;
+    }
+
+    public StringProperty findResultProperty() {
+        return findResult;
+    }
+
+    public StringProperty findStatusProperty() {
+        return findStatus;
+    }
+
+    public StringProperty removeKeyProperty() {
+        return removeKey;
+    }
+
+    public StringProperty removeResultProperty() {
+        return removeResult;
+    }
+
+    public StringProperty removeStatusProperty() {
+        return removeStatus;
+    }
+
+    public String getAddStatus() {
+        return addStatus.get();
+    }
+
+    public String getFindResult() {
+        return findResult.get();
+    }
+
+    public String getFindStatus() {
+        return findStatus.get();
+    }
+
+    public String getRemoveResult() {
+        return removeResult.get();
+    }
+
+    public String getRemoveStatus() {
+        return removeStatus.get();
+    }
+
+    public void add() {
+        final int key = Integer.parseInt(addKey.get());
+        final String value = addValue.get();
+
+        storage.add(key, value);
+
+        addStatus.set(Status.SUCCESS.toString());
```

Comment:
But what if User wants to add different value for the same key?
In this case we would make him angry because we doesn't save his inputs for further processing - a bit annoying, I suppose

Reply:
I don't familiar with this kind of behavior. Because I will become angry if I should alway clean those fileds before inserting new value. 
But ok, it's user experience topic, so I have no objections.